features/26-mfa/api.md

```diff
@@ -0,0 +1,961 @@
+
+In the future, the below factors will be supported:
+
+- [WebAuthn with CTAP2 or FIDO U2F](https://fidoalliance.org/specifications/)
+
+The developer can configure the enforcement of MFA.
+
+The session contains how the user was authenticated.
+
+# Configuration
+
+```yaml
+mfa:
+  # If enabled is false, the MFA API is disabled. THe user cannot register authenticator.
+  # Default is false.
+  enabled: false
+  # Valid values are 'off', 'optional' or 'required'.
+  # Default is 'optional'.
+  #
+  # 'off' means MFA is never required during authentication.
+  #
+  # 'optional' means MFA is required during authentication when the user has at least 1 activated authenticator. This is like how other major online services such as Google offer MFA.
+  #
+  # 'required' means MFA is always required during authentication.
+  enforcement: 'optional'
+  # The total maximum number of authenticator the user can register.
+  # Valid values is [0, 15].
+  # Default is 1.
+  maximum: 1
+  totp:
+    # The maximum number of TOTP authenticator the user can register.
+    # Valid values is [0, 5].
+    # Default is 0.
+    maximum: 0
+  oob:
+    sms:
+      # The maximum number of OOB-SMS authenticator the user can register.
+      # Valid values is [0, 5].
+      # Default is 0.
+      maximum: 0
+    email:
+      # The maximum number of OOB-Email authenticator the user can register.
+      # Valid values is [0, 5].
+      # Default is 0.
+      maximum: 0
+  bearer_token:
+    # How many days the bearer token is valid. Default is 30.
+    expire_in_days: 30
+  recovery_code:
+    # How many recovery code should be generated. Default is 16.
+    count: 16
+    # Enable listing the existing recovery code.
+    # Some services like GitHub allow listing the existing recovery code.
+    # By default listing is disabled, the user must regenerate a new set of
+    # recovery code.
+    list_enabled: false
+```
+
+# Configuration Recipes
+
+## At most 1 TOTP or 1 OOB-SMS or 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP and 1 OOB-SMS and 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  maximum: 3
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+```
+
+## At most 1 OOB-SMS
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  oob:
+    sms:
+      maximum: 1
+```
+
+# Endpoints
+
+## POST /mfa/totp/new
+
+### Request body
+
+```JSON
+{
+  "display_name": "my TOTP"
+}
+```
+
+- `display_name`: The display name of the new TOTP authenticator.
+- `authn_session_token`: Required only when there is no access token.
+
+### Response body
+
+```JSON
+{
+  "authenticator_id": "xxx",
+  "authenticator_type": "totp",
+  "secret": "secret"
```

Comment:
in real world use cases, is it easier if user provide a optional `label` and `issuer` to us, we can respond the QR Code image URL at once here?

Reply:
Your two comments are about the convenience of accepting either `otpauth_uri` or `(accountName, issuer)`. I agree it is more ergonomic to have this convenience. Let me include in the spec and open an issue to implement it.